src/main/scala/rocket/ALU.scala

```diff
@@ -64,50 +80,97 @@ abstract class AbstractALU[T <: ALUFN](val aluFn: T)(implicit p: Parameters) ext
+      val shin_hi_32 = Fill(32, isSub(io.fn) && io.in1(31))
       val shin_hi = Mux(io.dw === DW_64, io.in1(63,32), shin_hi_32)
       val shamt = Cat(io.in2(5) & (io.dw === DW_64), io.in2(4,0))
       (shamt, Cat(shin_hi, io.in1(31,0)))
     }
-  val shin = Mux(io.fn === aluFn.FN_SR  || io.fn === aluFn.FN_SRA, shin_r, Reverse(shin_r))
-  val shout_r = (Cat(aluFn.isSub(io.fn) & shin(xLen-1), shin).asSInt >> shamt)(xLen-1,0)
+  val shin = Mux(shiftReverse(io.fn), Reverse(shin_r), shin_r)
+  val shout_r = (Cat(isSub(io.fn) & shin(xLen-1), shin).asSInt >> shamt)(xLen-1,0)
   val shout_l = Reverse(shout_r)
-  val shout = Mux(io.fn === aluFn.FN_SR || io.fn === aluFn.FN_SRA, shout_r, 0.U) |
-              Mux(io.fn === aluFn.FN_SL,                           shout_l, 0.U)
+  val shout = Mux(io.fn === FN_SR || io.fn === FN_SRA || io.fn === FN_BEXT, shout_r, 0.U) |
+              Mux(io.fn === FN_SL,                                          shout_l, 0.U)
 
   // CZEQZ, CZNEZ
   val in2_not_zero = io.in2.orR
   val cond_out = Option.when(usingConditionalZero)(
-    Mux((io.fn === aluFn.FN_CZEQZ && in2_not_zero) || (io.fn === aluFn.FN_CZNEZ && !in2_not_zero), io.in1, 0.U)
+    Mux((io.fn === FN_CZEQZ && in2_not_zero) || (io.fn === FN_CZNEZ && !in2_not_zero), io.in1, 0.U)
   )
 
   // AND, OR, XOR
-  val logic = Mux(io.fn === aluFn.FN_XOR || io.fn === aluFn.FN_OR, in1_xor_in2, 0.U) |
-              Mux(io.fn === aluFn.FN_OR || io.fn === aluFn.FN_AND, io.in1 & io.in2, 0.U)
+  val logic = Mux(io.fn === FN_XOR || io.fn === FN_OR || io.fn === FN_ORN || io.fn === FN_XNOR, in1_xor_in2, 0.U) |
+              Mux(io.fn === FN_OR || io.fn === FN_AND || io.fn === FN_ORN || io.fn === FN_ANDN, in1_and_in2, 0.U)
 
-  val shift_logic = (aluFn.isCmp (io.fn) && slt) | logic | shout
+  val bext_mask = Mux(coreParams.useZbs.B && io.fn === FN_BEXT, 1.U, ~(0.U(xLen.W)))
+  val shift_logic = (isCmp (io.fn) && slt) | logic | (shout & bext_mask)
   val shift_logic_cond = cond_out match {
     case Some(co) => shift_logic | co
     case _ => shift_logic
   }
-  val out = Mux(io.fn === aluFn.FN_ADD || io.fn === aluFn.FN_SUB, io.adder_out, shift_logic_cond)
+
+  // CLZ, CTZ, CPOP
+  val tz_in = MuxLookup((io.dw === DW_32) ## !io.in2(0), 0.U)(Seq(
+    0.U -> io.in1,
+    1.U -> Reverse(io.in1),
+    2.U -> 1.U ## io.in1(31,0),
+    3.U -> 1.U ## Reverse(io.in1(31,0))
+  ))
+  val popc_in = Mux(io.in2(1),
+    Mux(io.dw === DW_32, io.in1(31,0), io.in1),
+    PriorityEncoderOH(1.U ## tz_in) - 1.U)(xLen-1,0)
+  val count = PopCount(popc_in)
+  val in1_bytes = io.in1.asTypeOf(Vec(xLen / 8, UInt(8.W)))
+  val orcb = VecInit(in1_bytes.map(b => Fill(8, b =/= 0.U))).asUInt
+  val rev8 = VecInit(in1_bytes.reverse).asUInt
+  val unary = MuxLookup(io.in2(11,0), count)(Seq(
+    0x287.U -> orcb,
+    (if (xLen == 32) 0x698 else 0x6b8).U -> rev8,
+    0x080.U -> io.in1(15,0),
+    0x604.U -> Fill(xLen-8, io.in1(7)) ## io.in1(7,0),
+    0x605.U -> Fill(xLen-16, io.in1(15)) ## io.in1(15,0)
+  ))
```

Comment:
RISC-V is elegant? :melting_face: 
For other future note, this is defined https://github.com/riscv/riscv-opcodes/blob/07b21cc5143a15959eda12e30aa40cea0971efe0/rv_zbb#L4-L6
Used for zvbb...

This remains some discussion on the where should unary decoding be? at DEC stage or at EXE stage?
For me I'd like to put it to dec, but the rocket-chip dec is hard to maintain...

Reply:
This should be done at DEC really, but I don't think it makes a big difference in the grand scheme